### PR TITLE
Hotfix Zstd breaking due to version change in runners

### PR DESCRIPTION
**Pull request by pdotl**
## Summary

Currently zstd check is dependent on the text output of the `zstd --version` to include `zstd command line interface`. This text changed to `*** Zstandard CLI (64-bit) v1.5.4, by Yann Collet ***` breaking the check and `@actions/cache` always falling back to `gzip` compression.

## Implementation

- Set to use `gzip` if the version output string is empty which should be the case if zstd is not installed. 
- Introduced `--quiet` flag to `zstd` so moving on the `zstd --version` should output machine readable version parsable via SemVer instead of text.
- Removed code that checks for version less than 1.3.2 as it was not working. Now we default to using zstd without `--long` as that is what currently happens.

## Next Steps

Will have another PR that fixes use of `zstd` with `--long` as intended. This release should unblock `zstd` usage.

## Issues

Fixes https://github.com/actions/toolkit/issues/1341

### CODE REVIEW

File: packages/cache/src/internal/cacheUtils.ts [hunk 1]
```diff
@@ -71,11 +71,13 @@ export async function unlinkFile(filePath: fs.PathLike): Promise<void> {
   return util.promisify(fs.unlink)(filePath)
 }
 
-async function getVersion(app: string): Promise<string> {
+async function getVersion(app: string, args?: string[]): Promise<string> {
```
Comment (bishal-pdMSFT) [inline packages/cache/src/internal/cacheUtils.ts:74]: As mentioned in another comment, this need not be an argument. Will simplify the code as well.

**Comment by bishal-pdMSFT** [comment]
Checks are failing. Please fix the unit tests.

File: packages/cache/src/internal/cacheUtils.ts [hunk 2]
```diff
@@ -94,18 +96,14 @@ async function getVersion(app: string): Promise<string> {
 
 // Use zstandard if possible to maximize cache performance
 export async function getCompressionMethod(): Promise<CompressionMethod> {
-  const versionOutput = await getVersion('zstd')
+  const versionOutput = await getVersion('zstd', ['--quiet'])
   const version = semver.clean(versionOutput)
 
-  if (!versionOutput.toLowerCase().includes('zstd command line interface')) {
+  if (versionOutput === '' || version === null) {
```
[the remark this answers is not in the document]
Comment (pdotl) [reply packages/cache/src/internal/cacheUtils.ts:102]: `versionOutput` is initialised with `''` but `version` returns null due to `semver.clean` returning null for versionOutput value that is not a valid semver.

[the remark this answers is not in the document]
Comment (pdotl) [reply packages/cache/src/internal/cacheUtils.ts:102]: Yes Bishal, I took a call to remove that from PR as it was currently redundant and we were not using `--long` at all. Since folks are having issues using `zstd` currently, I was thinking we can introduce `--long` after testing that it works but release now to unblock `zstd` usage. Although the change to fix `long` is also not much work.

File: packages/cache/src/internal/cacheUtils.ts [hunk 3]
```diff
@@ -94,18 +96,14 @@ async function getVersion(app: string): Promise<string> {
 
 // Use zstandard if possible to maximize cache performance
 export async function getCompressionMethod(): Promise<CompressionMethod> {
-  const versionOutput = await getVersion('zstd')
+  const versionOutput = await getVersion('zstd', ['--quiet'])
```
[the remark this answers is not in the document]
Comment (bishal-pdMSFT) [reply packages/cache/src/internal/cacheUtils.ts:99]: Got it

File: packages/cache/src/internal/cacheUtils.ts [hunk 4]
```diff
@@ -71,11 +71,13 @@ export async function unlinkFile(filePath: fs.PathLike): Promise<void> {
   return util.promisify(fs.unlink)(filePath)
 }
 
-async function getVersion(app: string): Promise<string> {
-  core.debug(`Checking ${app} --version`)
+async function getVersion(app: string, args?: string[]): Promise<string> {
```
Comment (bishal-pdMSFT) [inline packages/cache/src/internal/cacheUtils.ts:74]: Does using default value for args work? i.e. `args: string[] = []`. If so then we need not handle the undefined thing in L76

Comment (bishal-pdMSFT) [inline packages/cache/src/internal/cacheUtils.ts:74]: Name args as `additionalArgs` as the primay arg is `--version` which is not being passed.

File: packages/cache/src/internal/cacheUtils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bishal-pdMSFT) [reply packages/cache/src/internal/cacheUtils.ts:102]: Ok.

**Review by bishal-pdMSFT — APPROVED** [review]

File: packages/cache/src/internal/cacheUtils.ts [hunk 5]
```diff
@@ -71,11 +71,15 @@ export async function unlinkFile(filePath: fs.PathLike): Promise<void> {
...
+async function getVersion(
+  app: string,
+  additionalArgs: string[] = []
+): Promise<string> {
   let versionOutput = ''
+  additionalArgs.push('--version')
+  core.debug(`Checking ${app} ${additionalArgs.join(' ')}`)
   try {
-    await exec.exec(`${app} --version`, [], {
+    await exec.exec(`${app}`, additionalArgs, {
```
Comment (cdce8p) [inline packages/cache/src/internal/cacheUtils.ts:82]: Thanks again @pdotl!

Just saw this while looking at the PR. `'${app}'` seems a bit redundant. The first argument should be a string and `app` is already one. Maybe you can fix that when you address the `zstd --long` issue. It would be a bit much to open a separate PR for it 😅

**Review by Phantsure — COMMENTED** [review]
Looks fine to remove --long check entirely. --quiet would supress warnings but not clear how that would help in semver readable text

**Comment by cdce8p** [comment]
> --quiet would supress warnings but not clear how that would help in semver readable text

Found this in the documentation for `zstd`
> V, --version: display version number and exit. Advanced : -vV also displays supported formats. -vvV also displays POSIX support. -q will only display the version number, suitable for machine reading.

File: packages/cache/src/internal/cacheUtils.ts (hunk 5 above)
Comment (pdotl) [reply packages/cache/src/internal/cacheUtils.ts:82]: Sure @cdce8p will take this up in the next PR.
